drivers: dma: Add dts dma consumer support and generic DMA support for stm32

cybertale opened the pull request:
This PR adds driver support for DMA on f0/f1/f2/f3/f4/l0/l4 series stm32.
This should have been working on the f7 series SoCs, but there are some weird bugs inside the code that I cannot resolve.
There are two kinds of IP blocks are used across these stm32, one is the one that has been used on F2/F4/F7 series, and the other one is the one that has been used on F0/F1/F3/L0/L4 series.
    
Memory to memory transfer is only supported on the second DMA on F2/F4/F7 with 'st,mem2mem' to be declared in dts.
    
This driver depends on k_malloc to allocate memory for stream instances, so CONFIG_HEAP_MEM_POOL_SIZE must be big enough to hold them.
    
Common parts of the driver are in dma_stm32.c and SoC related parts are implemented in dma_stm32_v*.c.
    
This driver has been tested on multiple nucleo boards, including NUCLEO_F091RC/F103RB/F207ZG/F302R8/F401RE/L073RZ/L476RG with the loop_transfer and chan_blen_transfer test cases.

Fixes #9193

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

erwango commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -293,6 +293,14 @@
+			reg = <0x40020000 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x1>;
+			status = "disabled";
+			label = "DMA_0";
```
I'm curious about the label numeration here. Why not calling fist instance DMA_1?
Usually STM32 devices numeration starts from 1 to fit the reference manuals.

cybertale replied:
I was wanting to name them as DMA_1 and ADC_1 in my first place, while basically there are two reasons:

1. When I was writing applications, I found that CONFIG_DMA_0_NAME and CONFIG_ADC_0_NAME are defined in Kconfig rather than Kconfig.stm32. So DMA_0 is always the first DMA device no matter what driver we are using from kconfig. And the default content of CONFIG_DMA_0_NAME is 'DMA_0'. If we name it as DMA_1, one has to select the **second** DMA device in kconfig, which is DMA_1 to enable the **first** DMA device on STM32, which is DMA1. I think this may bring many bugs in the future.
2. Userspace applications can be shared between devices if all drivers use DMA_0 as its first device (What the OS wants).

cybertale replied:
I just noticed that all DMA are going to be enabled if DMA is selected, but ADC can be selected which to use. I think someday DMA will have this feature, too, to select which DMA device we want to enable.

erwango replied:
>     1. I found that CONFIG_DMA_0_NAME and CONFIG_ADC_0_NAME are defined in Kconfig rather

It was the same for other devices (UART/SPI/...) and it didn't prevent to set correct name for STM32 instances. Besides, once devices are fully moved to dts, CONFIG_DEV_X_NAME are no more required and deleted (cf serial driver example), so there is no more constraint on numbering.
Using UART_1 in zephyr when user looks for UART_2 in the datasheet is heavily confusing and should be avoided.

cybertale replied:
I agree to number it start from 1, and DMA driver should be like this.
While I think my problem was that we configure which serial to use in Kconfig.stm32, but configure which ADC to use in Kconfig under adc folder. It causes naming problems for ADC driver, though not related to DMA driver, ADC now has to be named from 0.

cybertale replied:
We have to merge the selection of which device to use into device's own kconfig for ADC I think, this is going to solve this problem in my head.

cybertale commented:
Hi @erwango , I have added dt support and tested with memcpy example on a f4 board, will add support for other socs later. Please have a look if you got sometime, mainly the dma_stm32.c. See if there is anything needs improving.

cybertale commented:
There are the major changes:
1. Adding dt support of course.
2. The original driver uses many unneccessary regs computation and passing, I replaced them with direct register access, so less code and cleaner.
3. Since it's for all stm32 socs and everyone of them may has different number of channels or DMA devices, max_streams is dynamically computed when init, by counting irq lines.

cybertale commented:
> @cybertale
> Very interesting work, but it broke the microphone support on ArgonKey.
> I think that we need to:
> 
>     1. change the i2s_ll_stm32.c because now some DMA info (like the name) are taken from
>        the DTS
> 
>     2. in board/96b_argonkey I guess that we need to declare dma_2 in DTS
> 
>     3. in sample for argonkey I had to define CONFIG_DMA_2 in prj.conf
> 
Sure, I noticed only one place and missed the others.
> 
> After having done all of this it refuse to compile because k_malloc() is missing.
> Can you help me?
Hi @avisconti , thanks for the interest, I forgot to mention that this driver depends on HEAP_MEM_POOL_SIZE to be not zero. Usually 512/1024 will work. 

This is because different series of STM32 has different number of DMA channels, and to only allocate channels we need, this driver uses a var 'max_streams' in the dma_stm32_data struct to count its how many channels do we have on the current instance, and uses k_malloc to allocate memory for these structures. 
In kconfig.stm32, I added 'depends on HEAP_MEM_POOL_SIZE != 0', but it seems like some other config options will set DMA_STM32 no matter what CONFIG_MEM_POOL_SIZE is. I think currently my solution is not good enough, the community may have some better ideas.

avisconti commented:
OK, now it compiles.
Unfortunately the flow is broken. I had similar problem in the past which I was able to solve
disabling completely the dma logs and increasing the dma interrupt priority. I think I need to spend some
time here.... :-(

cybertale commented:
@avisconti I have a f4 board with audio devices, I'll have a try today.

cybertale commented:
@avisconti Do you mean that the i2s interface is working but the playing is not continuous? Sorry I have never been used this interface before, it seems like to test play audio need many other support and there isn't a generic test application in zephyr. How can I help your debug?

avisconti commented:
> @avisconti Do you mean that the i2s interface is working but the playing is not continuous? Sorry I have never been used this interface before, it seems like to test play audio need many other support and there isn't a generic test application in zephyr. How can I help your debug?

No, actually the I2S+DMA is failing immediately. Yesterday evening I saw that DMA was failing with Fifo Error immediately. Are you based in Le Mans? If so you can get one ArgonKey board from Erwan and use my microphone sample.

cybertale commented:
@avisconti Unfortunately I'm in China, but since it fails from the beginning, I can try this driver with the configuration in i2s_ll_stm32.c to see if I can see the same error.

avisconti commented:
> @avisconti Unfortunately I'm in China, but since it fails from the beginning, I can try this driver with the configuration in i2s_ll_stm32.c to see if I can see the same error.

I'm debugging it and I actually found few mistakes.
I'll update you by end of my day

cybertale commented:
@avisconti Great! thx!

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,869 @@
+	}
+
+	if (id == data->max_streams) {
+		LOG_ERR("Unknow interrupt happened.");
```
we may want to return here, to avoid using a bad index.

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,869 @@
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC16);	\
+		break;							\
+	default:							\
```
the cases here should be 0,1,2,3 and not 1,4,8,16

cybertale replied:
This value represents how many data units does one burst send/receive, I don't understand why it should be 0, 1, 2, 3, comment in dma.h is blur about this. After searching 'source_burst_length' in Zephyr's tree, I see some usage in dma_qmsi.c, dma_sam_xdmac.c, dma_cavs.c, they treat this value with find_msb_set and something like that. This value should be 0 or 2^n I guess, right?

avisconti replied:
It seems again a mismatch like the data size. Before the driver was using dma_burst_index(), 
which was expecting the burst to be the exponent (i.e. 0 is 8-bit burst, 1 is 16-bit burst, ...).

But all the drivers seems to expect the burst size like 1, 4, 8, 16. So, we may change I2S driver
according to the API (even if in the burst case the dma.h seems not really clear in the 
explanation).

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,869 @@
+
+	m_size = table_m_size[index];
+	index = find_lsb_set(config->dest_data_size) - 1;
+	p_size = table_p_size[index];
```
Here m_size and p_size are wrong. For example in my case (I2S) I passed 1 for bot source and dest data size, because 
it is 16-bit transfer. But this code returned m_size = 0 and p_size = 0, which is wrong.
I remove the '- 1' to have index = 1 and it worked, but I'm not sure it is always correct.
Take a closer look here...

cybertale replied:
I'm a bit confused about this, m_size and p_size come from source_data_size and dest_data_size, which are width of source data/dest data (in bytes) as introduced in include/dma.h:121. So if we pass 1 for source_data_size and dest_data_size then m_size and p_size just should be 0 for representing one byte, right? In your case it is 16-bits, so we should pass 2 for 2 bytes to source_data_size/dest_data_size,  right?

cybertale replied:
i2s_ll_stm32.c:847 do gives source_data_size and dest_data_size 1 and commented them as '16 bit', but I believe this is the problem, they should be both assigned with 2, along with code in i2s_ll_stm32.c:865, do you think so?

avisconti replied:
I believe you are right. It seems that you have to specify the number of bytes (i.e 1 is 8-bits,
2 is 16-bits, 4 is 32-bits, 8 is 64-bits). I saw that all other DMAs drivers are doing that way.

I don't know why, but seems that before it was using dma_width_index() and in that case the
mapping was different (i.e. 0 was 8-bits, 1 was 16-bits, 2 was 32-bits and so on). But I tend
to think that this routine is not matching the API expectation.

So, yes, I think that we need to change i2s_ll_stm32.c driver to match source_data_size and 
dest_data_size (i.e. put 2 there to mean 16-bit).

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,869 @@
+		size_config = dma_stm32_width_config(config, true, dma, id);
+		LL_DMA_ConfigTransfer(dma, table_stream[id],
+				LL_DMA_DIRECTION_PERIPH_TO_MEMORY |
+				LL_DMA_MODE_CIRCULAR |
```
I had to remove CIRCULAR here, otherwise it was not working.
I guess that in this particular case (PERIPHERAL_TO_MEMORY) it does not apply.
Maybe.

cybertale replied:
Zephyr currently does not support to configure i DMA is in circular mode. I have not experienced with many DMA peripherals so I don't know if other DMA peripherals also have this feature. Before this feature is added, I'm thinking about configuring if it's in circular mode in dts. What do you think?

avisconti replied:
I don't know what @erwango has to say here. But yes, it seems to me that all the dma
configurations sooner or later has to be put in dts. So we may start with circular mode.

avisconti replied:
Here I have some doubts. We have two specific dts parts:

1. The dma controller section, which should describe the controller capabilities
2. The dma section inside a client driver (like I2S). Here we should describe what the client
    is using (channels, slot, direction,...)

Probably the CIRCULAR info goes here somehow.

cybertale replied:
Yes, back to Linux we have DMA controller part dts and DMA device part dts, and Zephyr supports this arch for SPI and I2C already, I think I can try to add a dma-devices.yaml to add support for this function.

avisconti replied:
Yes, seems the right direction

cybertale replied:
I'm working on the dma consumer in #15218 , But I have some doubts, some properties like if this channel is working in circular mode is not a generic property for all dma drivers. Only DMA of stm32 supports this functionality. I don't think we should put it in dts, but configure it in the controller side. There would be a property to specify which channel is in circular mode, like this:

dma: dma@40012000 {
...
        circular-channel = <0 2 3>;
...
}

Just because it's special I think, and dma consumer currently is not able to handle device related special properties. What do you think?

cybertale replied:
I just realized that this way low level dts's circular-channel would mistakely overwrite high level dts's circular-channel some time.

jli157 replied:
@cybertale Can you add CIRCULAR support and make it configurable? In my implementation #14916  for UART Async, the DMA is configured to use the circular mode ( which I extended).

cybertale replied:
Sure, will do it as soon as I got some time.

avisconti reviewed (commented):
With this changes I2S audio is now working fine.

avisconti commented:
Moreover I had to modify ArgonKey dts file (boards/arm/96b_argonkey/96b_argonkey.dts)
I added the following to generate the proper macro:

+
+&dma2 {
+       status = "ok";
+};

avisconti commented:
I had also to modify samples/boards/96b_argonkey/microphone/prj.conf

CONFIG_HEAP_MEM_POOL_SIZE=1024
CONFIG_DMA_2=y

avisconti commented:
Last thing (I was forgetting) I had to change i2s_ll_stm32 driver.
I had to  replace all CONFIG_DMA_2_NAME with DT_DMA_2_NAME.
Similarly, all CONFIG_DMA_1_NAME with DT_DMA_1_NAME.

cybertale commented:
Hi @avisconti , I added dma consumer support to the commits, please give it a test when you got some time, thx!
I also did some changes to the i2s driver. Old i2s driver assumes tx channel and rx channel use the same dma controller, and I changed it to be able to use different dma controllers because  from the dts perspective, we cannot be sure they are always the same dma controller, what do you think?

avisconti commented:
@cybertale 
First of all it looks like a very important piece of work. And it seems very well understood from
concept to implementation.

Unfortunately it does not compile on my side. So, I will try to give you my feedback in the next days.

I think that @erwango should also review the many variants included in this PR. I can do that for STM32F4.

cybertale commented:
Hi @avisconti, thanks, can you show me which part of the code is failing?

avisconti commented:
> Hi @avisconti, thanks, can you show me which part of the code is failing?

Now it is working perfectly!
I just changed a couple of things. I'll submit a review asap.

avisconti commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -1117,37 +1185,44 @@ struct queue_item rx_5_ring_buf[CONFIG_I2S_STM32_RX_BLOCK_COUNT + 1];
+			.dest_data_size = DT_I2S_5_DMA_DST_DATA_SIZE_RX,    /* 16bit default */
+			.source_burst_length = DT_I2S_5_DMA_SRC_BURST_LENGTH_RX, /* SINGLE transfer */
+			.dest_burst_length = DT_I2S_5_DMA_DST_BURST_LENGTH_RX,
+			.channel_priority = DT_I2S_5_DMA_PRIORITY_RX,
```
I had to add all of this stuff in soc/arm/st_stm32/stm32f1/dts_fixup.h
#define DT_I2S_5_DMA_CONTROLLER_TX          DT_ST_STM32_I2S_40015000_DMA_CONTROLLER_0
#define DT_I2S_5_DMA_CHANNEL_TX             DT_ST_STM32_I2S_40015000_DMA_CHANNEL_0
#define DT_I2S_5_DMA_SLOT_TX                DT_ST_STM32_I2S_40015000_DMA_SLOT_0
#define DT_I2S_5_DMA_SRC_DATA_SIZE_TX       DT_ST_STM32_I2S_40015000_DMA_SRC_DATA_SIZE_0
#define DT_I2S_5_DMA_DST_DATA_SIZE_TX       DT_ST_STM32_I2S_40015000_DMA_DST_DATA_SIZE_0
[...]

I did it only for DT_I2S_5 just to test my board. We need to add also the other stuff for all I2S controllers for all
variants. 

@erwango 
Do we really need to add this in dts_fixup or can we use in drivers the correct DTS generated names?

cybertale replied:
This as far as I know actually should be put in dts_fixup.h. I didn't add this because I thought there was only F4 support for previous DMA and I only added 5 I2S dts fixup defines to soc/arm/st_stm32/stm32f4/dts_fixup.h. Since other platforms need this, too, I'll add it later. 
Driver cannot use them directly because the register base address changes from device to device, in this circumstance it's 40015000, and for another it becomes 40016000 or something else.

avisconti replied:
> Driver cannot use them directly because the register base address changes from device to device, in this circumstance it's 40015000, and for another it becomes 40016000 or something else.

Correct.

avisconti replied:
For example, in case of LPS22HB sensor the dts scripts generate following macros:

#define DT_ST_STM32_I2C_V1_40005800_ST_LPS22HB_PRESS_5D_LABEL    "LPS22HB"
#define DT_ST_LPS22HB_PRESS_0_LABEL        DT_ST_STM32_I2C_V1_40005800_ST_LPS22HB_PRESS_5D_LABEL

And in the driver we no longer need to use dts_fixup, but just DT_ST_LPS22HB_PRESS_0_LABEL.

Here, maybe, we can still use dts_fixup file and let @erwango comment on this.

cybertale replied:
Sure, but I suspect Erwan is a bit busy these days, don't get too many responses from him, I can tell you what I'm thinking.
This is actually because you have only one LPS22HB sensor. The problem is if we enabled only one I2S sensor, no matter which one it is, dts is going to generate it as DT_ST_STM32_..._0_LABEL. This means it can represent I2S1, I2S2 or any one of them.
Like if we enabled I2S2 and I2S3, it will generate DT_ST_STM32_I2S_0_LABEL and DT_ST_STM32_I2S_1_LABEL to represent I2S2 and I2S3, so our driver would not be able to identify which device instance we are actually using. This depends on the logic of extract/reg.py. So base address like 40005800 and 40015000 and dts_fixups are always needed in our case, at least for the stm32 drivers I've been experienced.

erwango replied:
> And in the driver we no longer need to use dts_fixup, but just DT_ST_LPS22HB_PRESS_0_LABEL.

This is not clear yet how we'll use "indexed instances" (a continuous index starting from 0, generated at compilation depending on the enabled IP instances. Ex UART_1 UART_3 UART_5 will get _0, _1 and _2). Basically this works today where there is only one IP instance, sensors being an example.

So for now, dts_fixup is the way.

avisconti commented on drivers/dma/Kconfig.stm32:
```diff
@@ -0,0 +1,33 @@
+
+config DMA_1
+	bool "Enable STM32 DMA Controller 1"
+	default y
```
Why is it enabled by default? Is this correct?

If it is correct indeed we need then to disable it in samples/boards/96b_argonkey/microphone/prj.conf
CONFIG_DMA_1=n

cybertale replied:
This is something I talked about with Andrzej in #13501. We thought every driver should have at least one default instance to be enabled. Please have a look in that PR, I'm not sure it's the best way for every driver, this can be discussed.

avisconti replied:
I think it is not case for example for I2C and SPI. 
@erwango  any comment on this?
Anyway, in case you want to have a default we require to change the microphone sample
as I was saying above.

cybertale replied:
Makes sense, example's default configuration should be in example's configuration file.

erwango replied:
To me, DMA instance configuration is indeed board dependent. Instance we want to activate depends on what we want to do with the board and how the resources are shared between the various tasks.
DMA_1 enabled might not be relevant for all boards. And then if some instance is enabled is makes things complicated as each board should deactivate (or not) default instance and then enable the instance they need.
So it's better to keep it default n and provide configuration on a per board basis.

avisconti reviewed (changes requested):
As you can see there was only few things in order to have it working on ArgonKey board.
I didn't check all the variants. I think I can do it and @erwango can do it as well!

THanks for the huge work!

cybertale commented:
Hi @avisconti , @jli157 here are the changes I made you may concern:
1. Test cases i2s_api and i2s_speed are failing because HEAP_MEM_POOL_SIZE and DMA_2 are not configured, so I added 96b_argonkey.conf to both of them supplying HEAP_MEM_POOL_SIZE and moved DMA_2 into boards/arm/96b_argonkey/Kconfig.defconfig. 
2. I removed the default y for DMA_1.
3. I didn't add all the dts_fixups for all 8 series of STM32 but only 6 I2Ss of F4 series. Even though you mentioned 'soc/arm/st_stm32/stm32f1/dts_fixup.h', but I2S is only supported on F4 series in zephyr, so is this a typo issue which should have been 'stm32f4/dts_fixup.h'? Fix ups for other series should be added after support of their drivers are added I think. :confused:
4. I also added support for circular mode through setting both source_reload_en and dest_reload_en are true in dma_block_config.

cybertale commented:
Btw, According to STM32's reference manuals, circular mode will reset the address counter as what source_reload_end and dest_reload_en should mean, but transfer is not going to be stopped in stm32. API is not very clear about whether the transfer should be stopped. I think it should be stopped but these configurations are the only things we have related to circular mode and new api is usually hard to get merged, so hope this is fine.

jli157 commented:
> Hi @avisconti , @jli157 here are the changes I made you may concern:
> 
> 1. Test cases i2s_api and i2s_speed are failing because HEAP_MEM_POOL_SIZE and DMA_2 are not configured, so I added 96b_argonkey.conf to both of them supplying HEAP_MEM_POOL_SIZE and moved DMA_2 into boards/arm/96b_argonkey/Kconfig.defconfig.
> 2. I removed the default y for DMA_1.
> 3. I didn't add all the dts_fixups for all 8 series of STM32 but only 6 I2Ss of F4 series. Even though you mentioned 'soc/arm/st_stm32/stm32f1/dts_fixup.h', but I2S is only supported on F4 series in zephyr, so is this a typo issue which should have been 'stm32f4/dts_fixup.h'? Fix ups for other series should be added after support of their drivers are added I think. 😕
> 4. I also added support for circular mode through setting both source_reload_en and dest_reload_en are true in dma_block_config.

Thanks for your work! I'd like to find time to try the new changes with my UART async API implementation. 
Is it possible to figure out a way to statically allocate required memory rather than using dynamic memory allocation? Otherwise, you have to find a way to remind the developer of setting HEAP_MEM_POOL_SIZE > 0.

jli157 commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -0,0 +1,27 @@
+#define STM32_DMA_DST_INCREMENT           1
+
+#define STM32_DMA_NORMAL_MODE             0
+#define STM32_DMA_CIRCULAR_MODE           1
```
Actually, there is one more mode: Double Buffer mode, based on STM32F4's RM. Will you consider to add that one?

cybertale replied:
I would like to add it but this must be amended after the dma api supports it. And API is hard to change due to the stability requirement of zephyr. Maybe zephyr's dma api needs an application specified pointer pointing to a driver specified structure to convey some special configurations like if the dual buffer mode is used. :worried:

cybertale commented:
> 
> Thanks for your work! I'd like to find time to try the new changes with my UART async API implementation.
> Is it possible to figure out a way to statically allocate required memory rather than using dynamic memory allocation? Otherwise, you have to find a way to remind the developer of setting HEAP_MEM_POOL_SIZE > 0.

Different soc has different number of dma stream lines, and every stream instance would have to allocate some memory for each one of them. 
There is a way of allocating memory statically, which is giving how many stream lines this soc has in dts. Originally I thought we can just count how many interrupt lines to know how many stream lines do we have, but this can only be known in initialization code, so stream instances are allocated dynamically in initialization code.
The reason I do not like to give another property in dts is that it would be carrying some information that interrupts has already given, just not necessary I think. Not the best way of doing this. This PR is not on milestone so we should have enough time to think up a better approach.

erwango commented:
@cybertale , any status on this? 
Do you only require reviews?

erwango commented:
Well seems you need rebase first

erwango commented:
^^ @FRASTM

cybertale commented:
> @cybertale , any status on this?
> Do you only require reviews?

Yes, was just waiting for you and your colleague's reviews. I'll rebase this and test the code again.

erwango commented:
Well, there had been some delays. Things should move now. Can you rebase?

cybertale commented:
> Well, there had been some delays. Things should move now. Can you rebase?

Sure, it's too late here to finish it today, should be done tomorrow.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,952 @@
+	} else if (func_is_active_ht[id](dma)) {
+		/* Silently ignore spurious transfer half complete IRQ */
+		LOG_INF("Half of the transfer has been completed.");
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
```
Seems there are 2 different IPs to support in the driver.
In order to ease code reading, you can follow what has been done in SPI and I2C cases in which we declared Kconfig symbols that carry the IP variant information and then are selected by the series.
Cf  https://github.com/zephyrproject-rtos/zephyr/blob/528a2de29054c9234e19dd985f0fa8fb7dc05ee1/drivers/spi/Kconfig.stm32#L18
Youcan call them V1 and V2..
This way, when a new series require to support DMA, they only have to select the variant and we don't need to update the driver at each `#if define`

erwango replied:
Also, with this new symbol, you'll be able to use IS_ENABLED macro, which make things even easier to read.
You need to check how it works though.

cybertale replied:
I can do this, but it may cost some time to finish it.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,952 @@
+	case DMA_ADDR_ADJ_DECREMENT:
+		return -ENOTSUP;
+	default:
+	LOG_ERR("Memory increment error. %d", increment);
```
Wrong indentation here.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,952 @@
+		LOG_ERR("Source unit size error, %d",
+			config->source_data_size);
+		return -EINVAL;
+		}
```
Wrong indentation

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,952 @@
+		LOG_ERR("Dest unit size error, %d",
+			config->dest_data_size);
+		return -EINVAL;
+		}
```
Wrong indentation

erwango commented on dts/arm/st/f4/stm32f401.dtsi:
```diff
@@ -37,6 +37,16 @@
 			reg = <0x40003800 0x400>;
 			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00004000>;
 			interrupts = <36 5>;
+			dma = <&dma1 4 0 STM32_DMA_MEMORY_TO_PERIPH
```
Is it applicable to all boards that will use this SoC? Or is it board/application dependent?
In the latter case this should be done either at board level, either as an overlay in matching sample.

cybertale replied:
Which DMA instance and which channel to use is applicable to all boards using this SoC. And other configurations are depends on the I2S driver. I retrieved them from them old I2S driver when all of them are hard coded in the I2S driver. So at least every board using our stm32 i2s driver in zephyr should use this configuration.

cybertale replied:
Which DMA instance and which channel to use is applicable to all boards using this SoC. And other configurations are depends on the I2S driver. I retrieved them from them old I2S driver when all of them are hard coded in the I2S driver. So at least every board using our stm32 i2s driver in zephyr should use this configuration.

erwango replied:
@avisconti , can you comment here?
Is this configuration really SoC specific: applies for all socs (and socs derived from this one), on all boards, for all use cases.
For instance, if I don't use I2S driver, can't I reuse this channel for other purpose?

avisconti replied:
@erwango @cybertale 
Well, some time has passed by since I have explored the DMA so I need to reconstruct all the knowledge I had.
Anyway, every SoC defines its own table of dma-peripheral connections which is fixed of course, because it reflects the way the peripheral is connected to the DMA. So, for example, the DMA1 ch0 strm3 is connected to SPI2_Rx (I2S2_Rx) and DMA1 ch0 strm4 is connected to SPI2_Tx (I2S2_Tx).

So, the combination channel + stream is fixed to a particular peripheral. Here, it is defined for the I2S instance the DMA configuration.

avisconti replied:
I think this work is a huge tentative to move DMA information out from the I2S driver and move it to dts.
This can be potentially an example for other drivers too that wants to make use of DMA.

avisconti replied:
> Is it applicable to all boards that will use this SoC? Or is it board/application dependent?
> In the latter case this should be done either at board level, either as an overlay in matching sample.

The DMA to Peripheral connections are hardwired in the SoC and described in a table in SoC Reference Manual. So, it will be same for all boards using that SoC.

erwango replied:
ok, thanks for clarification. Closing this point.

erwango commented on dts/arm/st/f4/stm32f401.dtsi:
```diff
@@ -48,6 +58,16 @@
 			reg = <0x40003c00 0x400>;
 			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00008000>;
 			interrupts = <51 5>;
+			dma = <&dma1 5 0 STM32_DMA_MEMORY_TO_PERIPH
```
Ditto

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,952 @@
+};
+
+static u32_t table_channel[] = {
+	LL_DMA_CHANNEL_0,
```
Meanings of 'stream' and 'channel' between zephyr and different STM32 SoCs are different(Easy to confuse people). In this situation, LL_DMA_CHANNEL_n in some SoCs is the same thing as LL_DMA_STREAM_n in f2/4/7. And LL_DMA_CHANNEL_n in f2/4/7 is another story... Maybe I should add some comment here.

erwango replied:
My comment was specific to `LL_DMA_CHANNEL_0`.

cybertale replied:
Sorry didn't see the notification of your reply. I mean the arrays containing them have different names, one is called table_channel and the other one is called table_stream, that's why I have to copy the full table.

galak commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -0,0 +1,51 @@
+      description: required irq number and irq priorities
+      generation: define
+
+    support-m2m:
```
this looks like it should just be a boolean, ie if the property exists than its true, if not its false.

cybertale replied:
Right

cybertale commented:
@erwango Did some minor fixes you mentioned. It took some time to split the driver into common parts and SoC related parts. Common parts are now located in dma_stm32.c and SoC related parts are located in dma_stm32_v1.c and dma_stm32_v2.c.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,599 @@
+struct dma_stm32_config {
+	struct stm32_pclken pclken;
+	void (*config_irq)(struct device *dev);
+
```
Can you remove this line?

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,599 @@
+	u16_t count;
+	int ret;
+
+	if (id >= data->max_streams) {
```
For my information, for this function, have you considered use of `LL_DMA_Init` ?
If yes, what were the reasons why you chose not to use it?

cybertale replied:
No, didn't know this before. I can use LL_DMA_Init instead, it seems like lines of code they will use are almost the same.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,599 @@
+		if (!stm32_dma_disable_stream(dma, id)) {
+		}
+		/* After trying for 5 seconds, give up */
+		if (count++ > (5 * 1000) / 50) {
```
Not sure about the `/ 50` here

cybertale replied:
Right, this is a mistake

erwango commented on drivers/dma/dma_stm32_v1.c:
```diff
@@ -0,0 +1,301 @@
+	LL_DMA_CHANNEL_7,
+};
+
+void (*func_clear_ht[])(DMA_TypeDef *DMAx) = {
```
+1 about these tables.
About the naming, what about `func_ll_` to make it clear in the calling code ?

cybertale replied:
Sure

cybertale replied:
I will also rename table_stream and table_channel to table_ll_stream and table_ll_channel since they are also table of LL defines.

erwango commented on drivers/dma/dma_stm32_v1.c:
```diff
@@ -0,0 +1,301 @@
+	}
+}
+
+#define BURST_SIZE(dev, dev_short)					\
```
Not quite convinced about this trick.
It saves few lines (macro in only called once, so we save only one "duplication" ) but at reading comfort expense..
Can you revert to usual code?

cybertale replied:
Sure

erwango reviewed (changes requested):
Few change requested, otherwise, I'm pleased with the general driver structure.
About testing strategy, yu mention this has been tested on several boards. Was it against `test/drivers/dma` ?
If yes, can you update boards descriptions and enable dma on these boards so it can be regularly compiled and tested. If no, wwhat are the blockers?
In any case,  please edit PR description and make this clear.

cybertale commented:
> About testing strategy, yu mention this has been tested on several boards. Was it against `test/drivers/dma` ?

Sorry for the late reply, I was out for a competition last week. I was using the test case in tests/drivers/dma/loop_transfer to test the driver on all the boards.

erwango commented:
@cybertale,

> Sorry for the late reply, I was out for a competition last week. I was using the test case in tests/drivers/dma/loop_transfer to test the driver on all the boards.

Good! Then maybe you can update the tested boards yaml files sot they are tested in CI from now on?
Also, it'd be great to have this one merged in 2.0, and window freeze should happen end of next week. Do you think you can make the changes suggested by this date ?

cybertale commented:
> @cybertale,
> 
> > Sorry for the late reply, I was out for a competition last week. I was using the test case in tests/drivers/dma/loop_transfer to test the driver on all the boards.
> 
> Good! Then maybe you can update the tested boards yaml files sot they are tested in CI from now on?
> Also, it'd be great to have this one merged in 2.0, and window freeze should happen end of next week. Do you think you can make the changes suggested by this date ?

Only problem left is that this driver needs 'HEAP_MEM_POOL_SIZE' to be configured big enough in kconfig to make the test project work. Should be done today or tomorrow.

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,599 @@
+
+#define IRQ_INIT(dma, chan)                                             \
+do {									\
+	if (!irq_is_enabled(DT_DMA_##dma##_IRQ_##chan)) {		\
```
Yes, that's because some DMA channels on some SoCs share the same interrupt lines.

erwango replied:
ok, right, txs!

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,599 @@
+#define IRQ_INIT(dma, chan)                                             \
+do {									\
+	if (!irq_is_enabled(DT_DMA_##dma##_IRQ_##chan)) {		\
+		irq_connect_dynamic(DT_DMA_##dma##_IRQ_##chan,		\
```
This is also because different DMA channels are sharing the same interrupt lines, and static declaration will fail.

erwango replied:
ok, txs

galak commented on dts/bindings/dma/dma.yaml:
```diff
@@ -0,0 +1,37 @@
+      description: compatible strings
+      generation: define
+
+    label:
```
By including base.yaml, a lot of the duplication in here should go away.  Also we've removed 'generation:' field.

galak commented on scripts/dts/extract/dma.py:
```diff
@@ -0,0 +1,188 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
```
We've deprecated the dts/extract script.  We need to moved over to scripts/dts/edtlib.py and scripts/dts/gen_defines.py

cybertale replied:
:cry:

galak commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -0,0 +1,50 @@
+    compatible:
+      constraint: "st,stm32-dma"
+
+    reg:
```
by inherit from dma & base.yaml you can remove a lot of duplication here. (for things like reg).

galak commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -0,0 +1,50 @@
+      generation: define
+
+"#cells":
+  - channel
```
This feels like overkill and a fair amount of what should be specified in the API and not in .dts.

cybertale replied:
Sorry I don't quit get your point. 
Do you mean I shouldn't pass parameters in dts? These parameters are passed through dts to the dma consumer driver like i2s/adc/uart driver and they translate these parameters to the standard dma consumer API parameters.
Or you mean I should use dma API defined symbols like MEMORY_TO_MEMORY instead of STM32_DMA_MEMORY_TO_MEMORY defined in stm32_dma.h?

galak replied:
So the "parameters" should be what's needed to identify the dma channel, not all the runtime aspects of the dma transfer.  The application should be defining (direction, src_data_size, dst_data_size, src_burst_length, dst_burst_length, src_addr_increment, dst_addr_increment, circular_mode).  Think of the DTS as being HW description, not runtime parameters.

galak reviewed (changes requested):
A number of comments, at a high level we should be looking to follow the DTS specs on DMA here:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/dma/dma.txt

cybertale commented:
> A number of comments, at a high level we should be looking to follow the DTS specs on DMA here:
> 
> https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/dma/dma.txt

Thanks. Line 79 of this file says that

> but can contain any data that is required for configuring a channel.

Isn't parameters in dts/bindings/dma/st,stm32-dma.yaml are just doing this?

galak commented:
> Thanks.
> But line 79 of this file says that
> 
> > but can contain any data that is required for configuring a channel.
> 
> Isn't parameters in dts/bindings/dma/st,stm32-dma.yaml are just doing this?

See my comment inline in the binding review about the parameters.

We should probably utilize the description for st,stm32-dma from here:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/dma/stm32-dma.txt

cybertale commented:
@erwango Sorry to inform you that completing the changes galak asked for seems not possible be fore the end of this week for me. I need time to now how new scripts work. :cry:

galak commented:
> @erwango Sorry to inform you that completing the changes galak asked for seems not possible be fore the end of this week for me. I need time to now how new scripts work. 😢

No worries, we can work on getting this into 2.1

erwango commented:
@cybertale, any target for this work?
There are few others PR that rely on this one, such as #14916, should they wait?

cybertale commented:
> 
> 
> @cybertale, any target for this work?
> There are few others PR that rely on this one, such as #14916, should they wait?

Yes, I just started looking into the new code this week. I got it worked under the last rebase. One more week should be enough for me to port the code to new dts arch.

erwango commented:
@cybertale, fyi end of merge window for 2.1 is Nov 8th. Do you think you'll be able to match this time frame?

cybertale commented:
@erwango Yes, last 3 months was the time for Chinese students to find jobs before graduation, so I was a bit busy. Since the busy time is over I'm now able to put much more time on these open source  projects.

cybertale committed:
scripts: dts: add support for parsing stm32 dma consumer cells

Add support for parsing stm32 dma consumer cells, format of which
follows dma dts format declared in the Linux Kernel for the dma of
stm32:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/stm32-dma.txt

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
dts: arm: st: add dts support for DMA of stm32

Add dts support for f0/f1/f2/f3/f4/f7/l0/l4 series STM32.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: apply new dma consumer API to the stm32 i2s driver

This commit moves parameters previously hard coded in the driver to
device tree and apply the new dma consumer api to it.

This commit also adds support for passing address adjusment parameter
to dma in the i2s driver to make all parameters of dma consumers
functional.

The i2s driver assumes tx channel dma and rx channel dma depends on the
same dma controller, and this commit also change it to assume different
dma controllers can be used.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: dma: add generic driver support for all stm32

This commit adds driver support for DMA on f0/f1/f2/f3/f4/l0/l4
series stm32.
This should have been working on the f7 series SoCs, but there
are some weird bugs inside the code that I cannot resolve.

There are two kinds of IP blocks are used across these stm32, one is the
one that has been used on F2/F4/F7 series, and the other one is the one
that has been used on F0/F1/F3/L0/L4 series.

Memory to memory transfer is only supported on the second DMA on
F2/F4/F7 with 'st,mem2mem' to be declared in dts.

This driver depends on k_malloc to allocate memory for stream instances,
so CONFIG_HEAP_MEM_POOL_SIZE must be big enough to hold them.

Common parts of the driver are in dma_stm32.c and SoC related parts are
implemented in dma_stm32_v*.c.

This driver has been tested on multiple nucleo boards, including
NUCLEO_F091RC/F103RB/F207ZG/F302R8/F401RE/L073RZ/L476RG with the
loop_transfer and chan_blen_transfer test cases.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
boards: arm: 96b_argonkey: Update dma defconfig for tests

Previous defconfig for dma is DMA_STM32F4X in this board, while the new
generic driver uses DMA_STM32 to enable DMA support, and also dma driver
of stm32 now needs HEAP_MEM_POOL_SIZE to be big enough to hold dma
stream instances.

Additional .conf files are added for also adding HEAP_MEM_POOL_SIZE
configuration to two test cases.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
boards: 96b_stm32_sensor_mez: add dma1 to dts

i2s2 on this board depands on dma1, so add it to the dts file.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
boards: arm: add dma test support for some nucleo boards

Add DMA test support for 'loop_transfer', 'chan_blen_transfer' on
NUCLEO_F091RC/F103RB/F207ZG/F302RF8/F401RE/F746ZG/L073RZ/L476RG.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
CODEOWNERS: add myself as code onwer

Add myself as code owner of, drivers/dma/dma_stm32*, and
tests/drivers/i2s/*/96b_argonkey.conf.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

erwango commented on scripts/dts/gen_defines.py:
```diff
@@ -579,6 +583,46 @@ def write_clocks(node):
+            dma.priority = (dma.channel_config >> 16) & 0x3
+            dma.fifo_threshold = dma.features & 0x3
+
+            out_dev(node, "DMA_SLOT_{}".format(dma_i), dma.slot)
```
Feels strange to add compatible specific code at this point. 
Seems to me we should use binding "#cells" to generate this.
@ulfalizer, would you help sorting this out?

cybertale replied:
I personally also think so, while I didn't find a better way of doing this. There's seems like no examples of having such behaviour.

ulfalizer replied:
Yeah, would be nicer to use `#cells` here. See the comment I posted above.

Previously, new `phandle-array` properties had to be hardcoded in `edtlib.py`, because I hadn't figured out how to deal with them in a generic way. It was fixed by https://github.com/zephyrproject-rtos/zephyr/pull/19327 though.

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -837,37 +857,46 @@ struct queue_item rx_1_ring_buf[CONFIG_I2S_STM32_RX_BLOCK_COUNT + 1];
+			.channel_priority = DT_I2S_1_DMA_PRIORITY_RX,
 			.dma_callback = dma_rx_callback,
 		},
+		.src_addr_increment = DT_I2S_1_DMA_SRC_ADDR_INCREMENT_RX,
```
Trying to understand how this works globally.
From https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/stm32-dma.txt, I'd expect a mask is applied to get this info from bitmap

cybertale replied:
The bitmap is decoded in the gen_defines.py

erwango replied:
ok, got it, then we need maintainer point of view on this regard.
Not sure this is the right place.

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -0,0 +1,46 @@
+
+config DMA_STM32_V1
+	bool
+	depends on SOC_SERIES_STM32F2X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32F7X
```
Do we keep F7 for now?

cybertale replied:
I would like to keep it, I was thinking there maybe other people would like to help solving the buffer problem and even if the code does not work on f7 it does not breaks anything since there wasn't any support for the f7. But pushing code cannot pass the test is not ok, right? If it is not ok I'll remove this.

erwango commented on tests/drivers/dma/chan_blen_transfer/src/test_dma.c:
```diff
@@ -23,7 +23,18 @@
 #include <ztest.h>
 
+#if defined(CONFIG_SOC_FAMILY_STM32)
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
```
Would it make sense to use matching driver variant config symbol here (`CONFIG_DMA_STM32_Vx`)

cybertale replied:
Right

ulfalizer replied:
Unrelated nit, but could have `#ifdef CONFIG_FOO` instead of `#if defined(CONFIG_FOO)` when there's just one. I don't know if we just do it for consistency, but it looks a bit odd coming to Zephyr from the outside

cybertale replied:
> Unrelated nit, but could have `#ifdef CONFIG_FOO` instead of `#if defined(CONFIG_FOO)` when there's just one. I don't know if we just do it for consistency, but it looks a bit odd coming to Zephyr from the outside

No guidelines here so I think they are all ok. usually I would use what others are using. In Zephyr I saw many examples of using #if defined(CONFIG_FOO) so I'm using it. :thinking:

ulfalizer replied:
Think it's more that it's just getting copied around by people though. Fight the flow. ;)

erwango commented on tests/drivers/dma/loop_transfer/src/dma.c:
```diff
@@ -22,7 +22,17 @@
 static const char tx_data[] = "The quick brown fox jumps over the lazy dog";
 static char rx_data[TRANSFER_LOOPS][RX_BUFF_SIZE] = {{ 0 } };
 
+#if defined(CONFIG_SOC_FAMILY_STM32)
```
Same here ?

erwango reviewed (changes requested):
First batch of comments .
Let's get use of dts correct first (hope @galak and @ulfalizer can help in this regards)
and then we'll see on driver implementation details.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -579,6 +583,46 @@ def write_clocks(node):
                 controller.props["clock-frequency"].val)
 
 
+def write_dmas(node):
```
Would be a good idea to reuse the generic `write_phandle_val_list()` here for as much stuff as possible. Hopefully, the identifiers can fit into the same scheme used for other phandle arrays (`pwms`, etc.).

Try just calling just `write_phandle_val_list()` first, see what `#define`s you miss, and output those manually afterwards.

cybertale replied:
write_phandle_val_list() should be only outputting 'channel' 'slot' 'feature' and 'channel_configuration' here, which are declared in cells. I still have to decode 'periph_inc', 'mem_inc', 'periph_inc_fixed', 'priority' and 'fifo_threshold' here, right?

cybertale replied:
We have to get these vars from dts but for complying https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/stm32-dma.txt, I can only use 'channel', 'slot', 'channel_config' and 'features' in cells. write_phandle_val_list() will only save one line of code here.

ulfalizer replied:
Still better to use `write_phandle_val_list()` it for whatever it can be used for I think (and outputting whatever else needs to be output after calling it).

It also forces things like naming consistency, which is nice (`DMAS_CONTROLLER_...` instead of `DMA_CONTROLLER_...`, etc.) We have some inconsistency there for clocks already, which we want to get rid of.

ulfalizer commented on soc/arm/st_stm32/stm32f7/Kconfig.defconfig.series:
```diff
@@ -52,4 +52,11 @@ config ENTROPY_STM32_RNG
+config DMA_STM32_V1
+	default y
+
+endif # DMA
```
This is the same as this btw:

```
config DMA_STM32_V1
	default y if DMA_STM32
```

cybertale replied:
Ok, that's better

cybertale replied:
I think I would still use the style I was using, all other driver kconfig symbols in these files uses that style, should be consistent with them.

avisconti commented on dts/arm/st/f0/stm32f091.dtsi:
```diff
@@ -52,5 +52,21 @@
+			label = "DMA_2";
+		};
 	};
 };
```
Shouldn't this change (dma2) apply also to F051 and F071?

cybertale replied:
But F051 and F071 has only one DMA instance according to the datasheet.

avisconti replied:
Ok, correct.
I'm reading RM0091, and it seems infact that only STM32F091 has DMA2.
Sorry.

avisconti commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -311,6 +311,17 @@
+			#dma-cells = <4>;
+			reg = <0x40020000 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x1>;
+			interrupts = <9 0 10 0 10 0 11 0 11 0>;
```
Is it correct priority==0?
Moreover, how should I read the sequence of couples?

cybertale replied:
Priority is 0 means using the default priority I think. Sorry I don't understand what is 'sequence of couples', Could you give me a more specific description?

avisconti replied:
Like 9 0, 10 0, 10 0, 11 0, 11 0
Five couples which represent, I guess, irqs for the DMA channel/slot.
Can you describe it to me better? Thx

cybertale replied:
> Like 9 0, 10 0, 10 0, 11 0, 11 0
> Five couples which represent, I guess, irqs for the DMA channel/slot.
> Can you describe it to me better? Thx

Sorry for the late reply. For five couples, the first one in them represents the irq channel used in hardware, which is described in the datasheet, and the second one is the priority we would use. This DMA instance has 5 channels in hardware, so five couples of data. This is the default explanation for 'interrupts' for all dts nodes. Hope this anwsers your question.

avisconti commented:
@cybertale,
I can try to test it on ArgonKey, as I see a commit for I2S as well.

cybertale commented:
> @cybertale,
> I can try to test it on ArgonKey, as I see a commit for I2S as well.

Thanks! Was wanting you to help test in the first place. :smile:

avisconti commented:
> > @cybertale,
> > I can try to test it on ArgonKey, as I see a commit for I2S as well.
> 
> Thanks! Was wanting you to help test in the first place.

I decided to test it on x_nucleo_cca02m1 shield, as for the ArgonKey I lost my h/w setup. 
Cherry picking your 8 commits (from 46c1d86 to 5f3cc9f) I had to change DMA_STM32F4X into DMA_STM32. 
I have few warnings in dma_stm32.c:

/local/home/visconti/Projects/zephyrproject/zephyr/drivers/dma/dma_stm32.c:525:36: warning: 'dma_funcs' defined but not used [-Wunused-const-variable=]
 static const struct dma_driver_api dma_funcs = {
                                    ^~~~~~~~~
/local/home/visconti/Projects/zephyrproject/zephyr/drivers/dma/dma_stm32.c:494:12: warning: 'dma_stm32_init' defined but not used [-Wunused-function]
 static int dma_stm32_init(struct device *dev)
            ^~~~~~~~~~~~~~
/local/home/visconti/Projects/zephyrproject/zephyr/drivers/dma/dma_stm32.c:75:13: warning: 'dma_stm32_irq_handler' defined but not used [-Wunused-function]
 static void dma_stm32_irq_handler(void *arg)

And it doesn't work, so I have to debug tomorrow.

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,616 @@
+	default:
+		return -EINVAL;
+	}
+
```
I think that here you need to set the size, isn't it?
Like:

LL_DMA_SetDataLength(dma, table_ll_stream[id], size);

avisconti replied:
@cybertale 
I am now able to compile correctly, and I test it using I2S audio with microphone board.
But it does not work.
I tried to fix the dma_reload() function (see my latest review), but still does not work.
It seems that it stops receving data. Tomorrow I'll debug more.

galak commented:
Can we split this PR so the DTS changes are one PR and the driver changes are another.

avisconti commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -837,37 +857,46 @@ struct queue_item rx_1_ring_buf[CONFIG_I2S_STM32_RX_BLOCK_COUNT + 1];
+			.dma_slot = DT_I2S_1_DMA_SLOT_RX,
 			.channel_direction = PERIPHERAL_TO_MEMORY,
 			.source_data_size = 1,  /* 16bit default */
 			.dest_data_size = 1,    /* 16bit default */
```
I think that .source_data_size and .dest_data_size should be 2 (16bit == 2 bytes)

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,616 @@
+			config->dma_slot = 0;
+		}
+	}
+	DMA_InitStruct.Channel = config->dma_slot;
```
I think this is wrong.
DMA_InitStruct.Channel should be assigned with LL_DMA_CHANNEL_0/LL_DMA_CHANNEL_1/LL_DMA_CHANNEL_2/...

I noticed infact that the CR content was wrong here.
To fix it I added following code:

 >       switch (config->dma_slot) {
 >       case 0:
 >       >       DMA_InitStruct.Channel = LL_DMA_CHANNEL_0;
 >       >       break;
 >       case 1:
 >       >       DMA_InitStruct.Channel = LL_DMA_CHANNEL_1;
 >       >       break;
 >       case 2:
 >       >       DMA_InitStruct.Channel = LL_DMA_CHANNEL_2;
 >       >       break;
 [...]
 >      }

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,616 @@
+	}
+	DMA_InitStruct.Channel = config->dma_slot;
+
+	stm32_dma_get_fifo_threshold(config->head_block->fifo_mode_control);
```
I think you should assign the return value in DMA_InitStruct.FIFOThreshold.

avisconti replied:
Like:

>
> DMA_InitStruct.FIFOThreshold = stm32_dma_get_fifo_threshold(config->head_block->fifo_mode_control);
>

avisconti commented:
@cybertale @erwango 
With the modifications reported above, especially source and data size as well as the dma_slot assignment, I was able to pass correctly the peripheral-to-memory case on both x_nucleo_cca02m1 shield and ArgonKey.

avisconti commented:
@cybertale 
I pushed my test branch just in case:
https://github.com/avisconti/zephyr/pull/new/shield-microphones-dmanew

cybertale commented:
> @cybertale
> I pushed my test branch just in case:
> https://github.com/avisconti/zephyr/pull/new/shield-microphones-dmanew

Many thanks! I fixed the problems in the DMA driver. But I think the problem in the I2S driver is not related to the DMA driver and should be fixed in a septate PR. You want to push the fix to the I2S yourself or I help to do it?

avisconti commented:
> > @cybertale
> > I pushed my test branch just in case:
> > https://github.com/avisconti/zephyr/pull/new/shield-microphones-dmanew
> 
> Many thanks! I fixed the problems in the DMA driver. But I think the problem in the I2S driver is not related to the DMA driver and should be fixed in a septate PR. You want to push the fix to the I2S yourself or I help to do it?

@cybertale 
If it is not a problem and you agree with the changes you can do it. Then, I'll test again.

cybertale commented:
> > > @cybertale
> > > I pushed my test branch just in case:
> > > https://github.com/avisconti/zephyr/pull/new/shield-microphones-dmanew
> > 
> > 
> > Many thanks! I fixed the problems in the DMA driver. But I think the problem in the I2S driver is not related to the DMA driver and should be fixed in a septate PR. You want to push the fix to the I2S yourself or I help to do it?
> 
> @cybertale
> If it is not a problem and you agree with the changes you can do it. Then, I'll test again.

Ok.

cybertale commented:
@galak Splitting the PR to #19711 and #19712 .

avisconti commented:
@cybertale 
So are you going to close this PR?
Are #19711 and #19712 replacing it?

cybertale commented:
@galak Splitting the PR to #19711 and #19712 .

> @cybertale
> So are you going to close this PR?
> Are #19711 and #19712 replacing it?

Yes. I'm dividing the big PR to dts changes and driver changes regarding @galak 's advice.
